### PR TITLE
feature: Add support for type definition

### PR DESCRIPTION
**Pull request by tgodzik**
This approach uses is a slight modification of definition provider, where we try to additionally find the type symbol of the definition symbol. It seems to work ok and doesn't require too many changes.

In some cases it might not be clear where to go for the type definition, but I think most cases covered by tests make sense. For example the type definition for type is the same as definition.

There is a small bug for named parameters in Scala 2, but that is unrelated to this PR and can be fixed later.

Fixes https://github.com/scalameta/metals/issues/4510

### CODE REVIEW

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala [hunk 1]
```diff
@@ -11,7 +11,33 @@ import org.eclipse.lsp4j.Location
...
   import compiler._
-  def definition(): DefinitionResult = {
+
+  def definition(): DefinitionResult =
+    definition(findTypeDef = false)
+
+  def typeDefinition(): DefinitionResult =
+    definition(findTypeDef = true)
+
+  private def typeSymbol(tree: Tree, pos: Position) = {
```
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala:21]: Main logic for finding type symbol is in this method.

File: mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala [hunk 2]
```diff
@@ -161,34 +161,4 @@ object HoverProvider:
     case _ =>
       ju.Optional.empty()
 
-  private def expandRangeToEnclosingApply(
```
Comment (tgodzik) [inline mtags/src/main/scala-3/scala/meta/internal/pc/HoverProvider.scala:164]: Moved to MtagsEntichments

File: mtags/src/main/scala-3/scala/meta/internal/pc/PcDefinitionProvider.scala [hunk 3]
```diff
@@ -58,8 +66,43 @@ class PcDefinitionProvider(
...
-    import indexed.ctx
-    MetalsInteractive.enclosingSymbols(path, pos, indexed) match
+    given ctx: Context = indexed.ctx
+    definitionsForSymbol(
+      MetalsInteractive.enclosingSymbols(path, pos, indexed),
+      pos,
+    )
+  end findDefinitions
+
+  private def findTypeDefinitions(
```
Comment (tgodzik) [inline mtags/src/main/scala-3/scala/meta/internal/pc/PcDefinitionProvider.scala:76]: Main part for finding type symbol, this time for Scala 3

**Comment by tgodzik** [comment]
I fixed the issues with named parameters in the second commit. It should not correctly find both type and normal definition in case of named parameters.

File: mtags/src/main/scala-3/scala/meta/internal/pc/MetalsInteractive.scala [hunk 4]
```diff
@@ -191,14 +191,9 @@ object MetalsInteractive:
           val sym = funSym.owner.info.member(name).symbol
           List((sym, sym.info))
         else
-          val classTree = funSym.topLevelClass.asClass.rootTree
```
Comment (tgodzik) [inline mtags/src/main/scala-3/scala/meta/internal/pc/MetalsInteractive.scala:194]: Looks like this was not needed and actually sometimes returned empty tree.

File: tests/cross/src/test/scala/tests/typedef/TypeDefinitionSuite.scala [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+       |}
+       |""".stripMargin,
+  )
+
+  check(
+    "tuple",
+    // assert we don't go to `Tuple2.scala`
+    """|
+       |object Main {
+       |  @@(1, 2)
```
Comment (dos65) [inline tests/cross/src/test/scala/tests/typedef/TypeDefinitionSuite.scala:129]: why not?

File: mtags/src/main/scala-3/scala/meta/internal/pc/PcDefinitionProvider.scala [hunk 6]
```diff
@@ -58,8 +66,43 @@ class PcDefinitionProvider(
...
+      tree: Tree,
+      path: List[Tree],
+      pos: SourcePosition,
+      driver: InteractiveDriver,
+      indexed: IndexedContext,
+  ): DefinitionResult =
+    given ctx: Context = indexed.ctx
```
Comment (dos65) [inline mtags/src/main/scala-3/scala/meta/internal/pc/PcDefinitionProvider.scala:83]:
```suggestion
    import indexed.ctx
```

**Review by dos65 — COMMENTED** [review]
Nice work!

I left only few minor comments

File: tests/cross/src/test/scala/tests/typedef/TypeDefinitionSuite.scala (hunk 5 above)
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/typedef/TypeDefinitionSuite.scala:129]: Hmm.... that's currently done for go to definition, so the assumption was that type definition would do the same (I copied the tests from previous attempts). Essentially, this should be handled by go to definition if ever. 

We don't handle tuples anywhere unless they are explicitly used.

File: mtags/src/main/scala-2/scala/meta/internal/pc/completions/Completions.scala [hunk 7]
```diff
@@ -684,6 +684,9 @@ trait Completions { this: MetalsGlobal =>
         t match {
           // new User(age = 42, name = "") becomes transparent, which doesn't happen with normal methods
           case Apply(Select(_: New, _), _) => true
+          // for named args apply beceomes transparent but fun doesn't
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/pc/completions/Completions.scala]: Fixed!

**Review by dos65 — APPROVED** [review]
LGTM

**Comment by antosha417** [comment]
Hey @tgodzik! I'm very exited about this feature, nice job! 

Is it possible to extend it to return multiple locations for situations like 
```scala
val bar: Future[List[String]] = ???
val baz: Map[Int, String] = ???
```
I feel like sometimes I would want to go to `String` in such cases. 

Do you have any ideas on how to implement it?

**Comment by tgodzik** [comment]
We can for sure return multiple locations from definition. It should be possible to gather the symbols and return altogether.

We would need to return multiple symbols here https://github.com/scalameta/metals/blob/main/mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala#L21

and for Scala 3 around:
https://github.com/scalameta/metals/blob/main/mtags/src/main/scala-3/scala/meta/internal/pc/PcDefinitionProvider.scala#L85

And then combine all DefinitionResultImpl. We can have first symbol as the main one and all the rest just a simple locations. Though probably we could modify the DefinitionResult to have multiple symbols, but that might require changes in the entire codebase, so I would be fine with just having additional locations.

So the main task would be returning multiple symbols if the type is TypeApply and then combining them together.

**Pull request merged**